File: Python (Assignment 1.2 - Object-oriented)/BaseImage.py
from collections import namedtuple
from abc import ABC, abstractmethod


class BaseImage(ABC):
    class _Pixel:
        def __init__(self, red, green, blue):
            self.red = int(red)
            self.green = int(green)
            self.blue = int(blue)

        def __str__(self):
            return f'\n{self.red} {self.green} {self.blue}'

    _dimensions = namedtuple('dimensions', 'width height')

    def __init__(self, dimensions, themes, binary_format):
        self._Dimensions = BaseImage._dimensions(*dimensions)

        self._pixels = list()
        self._patterns = list()

        self._themes = themes

        self._binaryFormat = str(binary_format)

    @abstractmethod
    def read(self, file_name):
        pass

    @abstractmethod
    def draw(self, file_name):
        pass

    @abstractmethod
    def _choose_pattern(self):
        pass

    @abstractmethod
    def _write_to(self, file_name):
        pass

    def size(self):
        return self._Dimensions.width * self._Dimensions.height

File: C++ (Assignment 1.2 - Object-oriented)/BaseImage.cpp
#include <iostream>
#include <vector>

#include "BaseImage.h"
#include "PatternFactory.h"
#include "Pattern.h"

baseImage::baseImage(dimensions Dimensions, patternFactory* themes, std::string binaryFormat) 
	: pixels(Dimensions.height * Dimensions.width), Dimensions(Dimensions), themes(themes), binaryFormat(binaryFormat) {}

baseImage::~baseImage() {
	for (auto& pixel : pixels) { 
		delete pixel; 
		pixel = nullptr;
	}
	for (auto& pattern : patterns) { 
		delete pattern; 
		pattern = nullptr;
	}

	delete themes;
	themes = nullptr;
}

baseImage::pixel*& baseImage::operator[](int index) {
	return pixels[index];
}

const int baseImage::size() const { return pixels.size(); }

const baseImage::dimensions& baseImage::getDimensions() const  { return this->Dimensions; }
